### ISSUE TITLE
Automatically create input and output pins for object flow

### ISSUE THREAD

**Issue by sz332** [gaphor/gaphor#2483]
### Is your feature request related to a problem?

When working on an activity diagram, when connecting two actions it is very painful to always create the output and input pins.

### Describe the solution you'd like

- Have a global configuration: automatically create pins for object flows (default false)
- When connecting two activities and when this mode is set, automatically create the pins and make the object flow between the two pins.
- The pins shall be created around the location of the drawing of object flow.

**Comment by amolenaar** [comment]
>     * Have a global configuration: automatically create pins for object flows (default false)

Can't we just create a pin? We strive for as little configuration as possible and just do the right thing.

I agree that this could be helpful. It should not be too hard to implement.

**Comment by sz332** [comment]
I can totally live with the ability to automatically create the pin, and if it is not comfortable for someone, he can always just move the connection from the pin to the action.

**Comment by danyeaw** [comment]
It is completely valid SysML and UML to not use pins at all. This is especially very common when creating process diagrams.

**Comment by amolenaar** [comment]
But aren't those using Control Flow? For Object Flow we need ports.

**Comment by sz332** [comment]
I think we need to check what the UML specification actually says, but I have seen many cases when people just used object flows between actions without the use of the pins. It might or might not be valid, but would be interesting to see the reason behind that.

Because of that I would say that we need to find a way to "toggle" between those two functionalities:

- connect two actions with an object flow
- connect two actions with an object flow while creating pins

How exactly, I consider it an implementation question, but some ideas (which we can start with and improve on)
- Have a switch button in project configuration
- Have a button somewhere which switches between pin creation or not
- Just add another tool on the toolbar that says: create object flow with ports

**Comment by danyeaw** [comment]
What if we automatically create pins depending on the modeling being done:

1.  Automatic for Modeling System Behavior in Swimlanes
![image](https://github.com/gaphor/gaphor/assets/10014976/d71a440a-e6ce-4f19-a996-ddb9cfe43335)

2. Not-Automatic for modeling a Flow Chart outside of Swimlanes
![image](https://github.com/gaphor/gaphor/assets/10014976/cf162a22-c01a-49ce-8a29-bffbc40fa6de)

**Comment by sz332** [comment]
I like the direction, but can we say that whether we shall automatically create pins (only between actions!) or not is depending on the presence or absence of swimlanes? What do you think?

**Comment by marek-piirikivi** [comment]
I did some reading within UML specification and it proposes both versions - so pin's seems to be optional :thinking: So, it might be not possible to automatically enforce the pin creation if they are optional. However, what we can do is have some hotkey that can be done which would automatically create a pin for us. Another alternative would be to have a customizable settings.

**Comment by sz332** [comment]
Pins are optional but very often used, especially when modeling a software system. I would argue that this could be an optional feature, I could totally live with a separate tool/icon.

**Comment by marek-piirikivi** [comment]
@sz332 
I agree. I have never seen an object flow without pin. However; seems like the standard does not care. I have not checked SysML standard yet - it generally seems to be stricter about the things that are optional in UML; so we may still get it through the profile.

**Comment by marek-piirikivi** [comment]
I have a proposal.

Everything stays as is; but then you have just created a flow to the activity which is not connected to the pins; a small floating button appears near the connection end/beginning (whichever does not yet have a pin). When clicking this small button a pin is added. And if you ignore, it just goes away as soon as you do anything else.
